packages/account/src/BaseAccount.ts

```diff
@@ -0,0 +1,171 @@
+    provider!: JsonRpcProvider
+    entryPoint!: EntryPoint_v100
+    chainId!: ChainId
+    signer!: Signer
+
+    constructor(readonly smartAccountConfig: SmartAccountConfig) {
+        const { bundlerUrl, epAddress } = smartAccountConfig
+        if (bundlerUrl)
+            this.bundler = new Bundler({
+                bundlerUrl,
+                epAddress
+            })
+    }
+
+    async estimateUserOpGas(userOp: UserOperation): Promise<UserOperation> {
+        if (!userOp.callData)
+            throw new Error("calldata is not present for estimation")
+        if (!this.bundler) {
+            // if no bundler url is provided run offchain logic to assign following values of user
+            // maxFeePerGas, maxPriorityFeePerGas, verificationGasLimit, callGasLimit, preVerificationGas
+            const feeData = await this.provider.getFeeData()
+            if (EIP1559_UNSUPPORTED_NETWORKS.includes(this.chainId)) {
+                // assign gasPrice to both maxFeePerGas and maxPriorityFeePerGas in case chain does not support Type2 transaction
+                userOp.maxFeePerGas = userOp.maxPriorityFeePerGas = feeData.gasPrice ?? await this.provider.getGasPrice()
+            } else {
+                if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas)
+                    throw new Error("Unable to get maxFeePerGas and maxPriorityFeePerGas from provider")
+                userOp.maxFeePerGas = feeData.maxFeePerGas
+                userOp.maxPriorityFeePerGas = feeData.maxPriorityFeePerGas
+            }
+            userOp.verificationGasLimit = await this.getVerificationGasLimit()
+            userOp.callGasLimit = await this.provider.estimateGas({
+                from: this.smartAccountConfig.epAddress,
+                to: userOp.sender,
+                data: userOp.callData
+            })
+            userOp.preVerificationGas = this.getPreVerificationGas(userOp)
+        }
+        const userOpGasEstimatesResonse = await this.bundler.getUserOpGasFields(userOp, this.chainId)
+        console.log('userOpGasEstimatesResonse ', userOpGasEstimatesResonse);
+        const { callGasLimit, verificationGasLimit, preVerificationGas, gasPrice, maxFeePerGas, maxPriorityFeePerGas } = userOpGasEstimatesResonse.result
+        if (gasPrice)
+            userOp.maxFeePerGas = userOp.maxPriorityFeePerGas = gasPrice
+        else {
+            userOp.maxFeePerGas = maxFeePerGas
+            userOp.maxPriorityFeePerGas = maxPriorityFeePerGas
+        }
+        userOp.verificationGasLimit = verificationGasLimit
+        userOp.callGasLimit = callGasLimit
+        userOp.preVerificationGas = preVerificationGas
+        return userOp
+    }
+
+    async getPaymasterAndData(userOp: UserOperation): Promise<UserOperation> {
+        console.log('paymaster state ', this.paymaster);
+        if (this.paymaster) {
+            userOp.paymasterAndData = await this.paymaster.getPaymasterAndData(userOp)
+        }
+        return userOp
```

Comment:
user won't call this directly. I am calling it inside buildUserOp function. so it paymaster instance is define. It will attach paymasterAndData by calling paymaster service

Reply:
But if paymaster instance is not defined, it'll return empty value. How will we handle situations where user is expecting a paymasterAndData but we are returning empty?